rest_framework/permissions.py

```diff
@@ -51,24 +52,45 @@ class AND:
     def __init__(self, op1, op2):
         self.op1 = op1
         self.op2 = op2
+        self.message1 = getattr(op1, 'message', None)
+        self.message2 = getattr(op2, 'message', None)
+        self.message = None
 
     def has_permission(self, request, view):
-        return (
-            self.op1.has_permission(request, view) and
-            self.op2.has_permission(request, view)
-        )
+        if not self.op1.has_permission(request, view):
+            self.message = self.message1
+            return False
+
+        if not self.op2.has_permission(request, view):
+            self.message = self.message2
+            return False
+
+        return True
 
     def has_object_permission(self, request, view, obj):
-        return (
-            self.op1.has_object_permission(request, view, obj) and
-            self.op2.has_object_permission(request, view, obj)
-        )
+        if not self.op1.has_object_permission(request, view, obj):
+            self.message = self.message1
+            return False
+
+        if not self.op2.has_object_permission(request, view, obj):
+            self.message = self.message2
+            return False
+
+        return True
```

Comment:
A little shorter and suddenly a message will be split for `has_permission` and `has_object_permission` methods:
```python
class AND:
    def __init__(self, op1, op2):
        self.op1 = op1
        self.op2 = op2

    def has_permission(self, request, view):
        if not self.op1.has_permission(request, view):
            self.message = getattr(op1, 'message', None)
            return False

        if not self.op2.has_permission(request, view):
            self.message = getattr(op2, 'message', None)
            return False

        return True

    def has_object_permission(self, request, view, obj):
        if not self.op1.has_object_permission(request, view, obj):
            self.message = getattr(op1, 'message', None)
            return False

        if not self.op2.has_object_permission(request, view, obj):
            self.message = getattr(op2, 'message', None)
            return False

        return True
```